### ISSUE TITLE
AWS S3 SDK does not work on Unity 2019.1 Android

### ISSUE THREAD

**Issue by Troshchuk** [aws/aws-sdk-net#1262]
AWS S3 SDK does not work on Unity 2019.1

## Expected Behavior
AWS S3 SDK works on Unity 2019.1

## Current Behavior
After update to Unity 2019.1 AWS s3 sdk does not work. After call `UnityInitializer.AttachToGameObject(this.gameObject);` I see in logs exception:

```
04-17 22:51:29.910 4895-4917/? E/Unity: ArgumentException: Object of type 'System.Object[]' cannot be converted to type 'UnityEngine.AndroidJavaObject[]'.
                                          at System.RuntimeType.CheckValue (System.Object value, System.Reflection.Binder binder, System.Globalization.CultureInfo culture, System.Reflection.BindingFlags invokeAttr) [0x00000] in <00000000000000000000000000000000>:0 
                                          at System.Reflection.MonoMethod.ConvertValues (System.Reflection.Binder binder, System.Object[] args, System.Reflection.ParameterInfo[] pinfo, System.Globalization.CultureInfo culture, System.Reflection.BindingFlags invokeAttr) [0x00000] in <00000000000000000000000000000000>:0 
                                          at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <00000000000000000000000000000000>:0 
                                          at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0 
                                          a
```

## Your Environment
* AWSSDK.Core version used: AWSSDK.S3.3.3.101.4
* Operating System and version: Android
* Targeted .NET platform: 4.x

**Comment by sstevenkang** [comment]
How are you targetting .NET 4.X with unity?  My understanding is that Unity is only compatible with netstandard2.0 and PCL.  Could you take a look at the dll that your application is loading and identify which platform Unity is resolving?

**Comment by Troshchuk** [comment]
Regarding .NET 4.x - https://docs.microsoft.com/en-us/visualstudio/cross-platform/unity-scripting-upgrade?view=vs-2019

**Comment by Troshchuk** [comment]
Any platform

**Comment by Troshchuk** [comment]
With Unity 2018.3.1f1 it works fine. But after updating of Unity to 2019.1.0f2 it stopped working

**Comment by sstevenkang** [comment]
Are you using NuGet package to resolve your dependencies?  If so, could you take a peek at the DLL(such as dotpeek) and make sure it's either BCL45 or netstandard2.0 variant?   I am guessing it's resolving to PCL dll.

**Comment by Troshchuk** [comment]
No, I don't. Just downloaded and import to Unity from https://github.com/aws/aws-sdk-net/blob/master/Unity.README.md -> http://sdk-for-net.amazonwebservices.com/latest/aws-sdk-unity.zip

**Comment by sstevenkang** [comment]
Unity 2018.1 and going forward should support netstandard2.0 out of the box.  Instead of using the unity SDK build, could you try taking dependency on  NuGet packages?  Are there APIs in the unity flavor of the SDK that's not in vanilla AWS SDK you need?

**Comment by Troshchuk** [comment]
I added NuGet for Unity (https://github.com/GlitchEnzo/NuGetForUnity), installed AWSSDK.Core.3.3.100.7 and AWSSDK.S3.3.3.101.5 (latest from NuGet) and adopted code. 
When called `s3Client = new AmazonS3Client(accesKey, secretKey, RegionEndpoint.EUWest1); ` throws ``` 04-19 18:16:02.375 11494-11514/? E/Unity: NotSupportedException: System.Configuration.ConfigurationManager::get_AppSettings
                                            at System.Configuration.ConfigurationManager.get_AppSettings () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Amazon.AWSConfigs.GetConfig (System.String name) [0x00000] in <00000000000000000000000000000000>:0 
                                            at Amazon.AWSConfigs..cctor () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Amazon.Runtime.ClientConfig..ctor () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Amazon.S3.AmazonS3Config..ctor () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Amazon.S3.AmazonS3Client..ctor (Amazon.Runtime.AWSCredentials credentials, Amazon.RegionEndpoint region) [0x00000] in <00000000000000000000000000000000>:0 
                                            at RemoteContent.InitS3 () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Game.InitRemoteContent () [0x00000] in <00000000000000000000000000000000>:0 
                                            at Game+<Init>d__40.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 
                                            at UnityEngine.SetupCoroutine.InvokeMoveNe```
I think not just the unity package created. I see specific `AWSConfigs.unity.cs` configuration for Unity (https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Core/_unity/AWSConfigs.unity.cs)

**Comment by Troshchuk** [comment]
I tried to add System.Configuration.ConfigurationManager from NuGet but exception was the same

**Comment by sstevenkang** [comment]
It looks like this is an issue of how NuGet incorrectly resolves DLLs.  Could you try one more thing?  Could you try unpacking the NuGet package(it's just a zip file) and grab the dlls under `netstanard2.0` ?  Reference those DLLs directly?

**Comment by Troshchuk** [comment]
Packages are already unpacked. 
AWSSDK.Core.3.3.100.7/lib/net45/AWSSDK.Core
AWSSDK.S3.3.3.101.5/lib/net45/AWSSDK.S3

**Comment by Troshchuk** [comment]
There are not any nestandard2.0 folders

**Comment by Troshchuk** [comment]
I found `AWSSDK.Core.3.3.100.7.nupkg`, renamed it to `AWSSDK.Core.3.3.100.7.zip` and unpacked. There are a lot of folders in `lib` folders and `nestandard2.0` folder also exists

**Comment by Troshchuk** [comment]
I removed `net45` and paste `nestandard2.0` and it works fine. 
@sstevenkang Thanks you very much. 
But I think you should simplify installing sdk for Unity users.

**Comment by sstevenkang** [comment]
Your feedback is totally warranted. Thanks for doing some investigating with us.  We plan on sharing news regarding how we are going to better support Unity going forward.

**Comment by asoule** [comment]
@sstevenkang The solution here is a little unclear to me.  Just so I'm following correctly, the proposed solution is to depend on the nuget package instead of the unitypackage, and then open said package as a zip file, and remove a directory in lib called net45. Is this correct?

Also, I would be interested in learning how to achieve the same goal using your GitHub repo, as a solution that is a little easier to maintain long-term.

**Comment by JakobJJ** [comment]
I'm also facing the same issues and wondering over the solution. @Troshchuk Maybe you can give us a a little more detailed steps for solving the issue?

@sstevenkang And I am also interested achieve the same goal using your GitHub repo.

**Comment by matteo-prosperi** [comment]
Unfortunately nuget doesn't allow to consistently choose a specifc target framework (in this case `nestandard2.0`) out of packages that contains multiple versions of the libraries.

The workaround we currently suggest is to go to the nuget page for the SDK packages (https://www.nuget.org/packages/AWSSDK.S3, https://www.nuget.org/packages/AWSSDK.Core), download the packages and open them as .zip files. You can extract the content of the `lib\netstandard2.0\` folder of the packages you plan to use (always including AWSSDK.Core) and then directly reference the extracted DLLs from your project. Because the SDK DLLs target NetStandard 2.0 you can use them across multiple platforms (https://docs.microsoft.com/en-us/dotnet/standard/net-standard).

We are currently investigating ways improve the experience for Unity users.

**Comment by JakobJJ** [comment]
Sorry for a late response. I have just had the time to test the solution. By using this solution there is issues like

error CS0103: The name 'UnityInitializer' does not exist in the current context
error CS0117: 'AWSConfigs' does not contain a definition for 'HttpClient'
error CS0117: 'AWSConfigs' does not contain a definition for 'HttpClientOption'

Also when will you be able to share some news for the Unity developers?

**Comment by Troshchuk** [comment]
Hi guys,

I added NuGet for Unity (https://github.com/GlitchEnzo/NuGetForUnity) and installed:
AWSSDK.Core
AWSSDK.S3
In `Packages` folder I found 2 folders with AWSSDK.Core.3.3.100.7.nupkg and AWSSDK.S3.3.3.101.5.nupkg . I unpacked them and copied `netstandard2.0` from here into `lib` folder. And removed `net45` from lib folder. And everything works fine

**Comment by JakobJJ** [comment]
@Troshchuk  thanks for the answer. But did you not get any errors regarding the AWSConfig? for me it seems like I get tons of errors.

**Comment by Troshchuk** [comment]
@JakobJJ When I tried to use AWSSDK with `net45` I had 
```NotSupportedException: System.Configuration.ConfigurationManager::get_AppSettings``` 
but after replacemnt to `netstandard2.0` everething is ok

**Comment by matteo-prosperi** [comment]
@JakobJJ Make sure you reference the `netstandard2.0` SDK assemblies and that you use the API reference at https://docs.aws.amazon.com/sdkfornet/v3/apidocs/Index.html. The `nestandard2.0` assemblies don't contain the `UnityInitializer` class and the `AWSConfigs` class doesn't contain the `HttpClient` or `HttpClientOptions` properties. As you can see [here](https://github.com/aws/aws-sdk-net/tree/master/sdk/src/Core/_unity), all of those are defined here in a unity-specific folder. If you are porting existing code using the Unity packages to the new netstandard2.0 assemblies, you will likely have to do some refactoring.

I am not able to comment on unannounced features but, if you have suggestions on what would make it easier to use the netstandard2.0 SDK in Unity, feel free to open an issue on this repo and tag me directly.

@Troshchuk @JakobJJ @asoule If you want a simpler way to get all the netstandard2.0 assemblies to use in your Unity project, you can use something like this [PowerShell script](https://github.com/aws/aws-sdk-net/files/3149987/DownloadSDK.zip). Obvioulsy you won't need all downloaded assemblies, just `AWSSDK.Core.dll` and those for the services you use (like `AWSSDK.S3.dll` for example).

**Comment by JakobJJ** [comment]
Yes then i see.. Thank you for clearing this up. I am porting a existing unity project that were using the old aws unitypackage. And that is why  I'm getting all of these issues. I will open an issue and post some suggestions. Thank you @matteo-prosperi

**Comment by asandu58** [comment]
Solved by exporting my project as a unity package and importing it into a new project made in Unity 2018.3.14f1...after that you only have to write your tags again and delete the following files: androidx.browser.browser-1.0.0 
androidx.browser.browser-1.0.0.aar 
androidx.core.core-1.0.0 
androidx.core.core-1.0.0.aar 
from \YoureProjectLocation\Assets\Plugins\Android

**Comment by eladfein** [comment]
> Yes then i see.. Thank you for clearing this up. I am porting a existing unity project that were using the old aws unitypackage. And that is why I'm getting all of these issues. I will open an issue and post some suggestions. Thank you @matteo-prosperi
@JakobJJ I am also trying to port my existing project, 
were you able to do it? 
if so can you please explain how?

**Comment by sahilgoyal123** [comment]
Not able to resolve it. I have downloaded the Nuget package, and converted it into zip. Now, If I delete AWSSDK.Core.dll from Unity, and copy paste the AWSSDKCore.dll from netstandard2.0 folder, then UnityInitializer file is not found. When should we expect a fix from AWS?

**Comment by matteo-prosperi** [comment]
@sahilgoyal123 This is expected. The AWS SDK for .NET is not API-compatible with the AWS Mobile SDK for Unity. You will have to rework some of your code in order to transition to the .NET Standard 2.0 libraries.

In case there are particular Unity-specific or mobile-specific features you are missing, please open an issue on this repo and we will consider the feature request for development.

**Comment by biemann** [comment]
I am having the same issue as the OP. I used the AWS Moble SDK for Unity.  Following the steps described above, I was able to make my code compile except the UnityInitializer. If I am using now the AWS SDK for .NET, do I still need to attach the Unity Initializer or can I make it work using the .NET methods? 
Refactoring the code to make the Unity Initializer work seems to be a tedious task.

**Comment by QuentinGprd** [comment]
> I am having the same issue as the OP. I used the AWS Moble SDK for Unity. Following the steps described above, I was able to make my code compile except the UnityInitializer. If I am using now the AWS SDK for .NET, do I still need to attach the Unity Initializer or can I make it work using the .NET methods?
> Refactoring the code to make the Unity Initializer work seems to be a tedious task.

For me using only AWS Simple Email Service, I didn't need a Unity Initializer.

However steps in the [Unity.README](https://github.com/aws/aws-sdk-net/blob/master/Unity.README.md#unity-sdk-fundamentals), with link.xml are still required.

**Comment by matteo-prosperi** [comment]
Hello,
in case you decide to use the .NET Standard 2.0 version of the AWS SDK for .NET we have made available a [download in .zip format](https://sdk-for-net.amazonwebservices.com/latest/v3/aws-sdk-netstandard2.0.zip) that should be more convenient for Unity users (we also have version specific downloads using a URI like [this one](https://sdk-for-net.amazonwebservices.com/releases/aws-sdk-netstandard2.0-3.3.590.0.zip)).
We have also continued our testing and, whenever using IL2CPP, we recommend to add a file named `link.xml` in your project assets and configure it so that IL2CPP doesn't strip any of the AWSSDK assemblies.
It should look like this:
```
<linker>
       <assembly fullname="AWSSDK.Core" preserve="all"/>
       <assembly fullname="AWSSDK.DynamoDBv2" preserve="all"/>
       <assembly fullname="AWSSDK.Lambda" preserve="all"/>
</linker>
```
Please make sure to include all the AWSSDK dlls that you are using.
The usage of `link.xml` is described [here](https://docs.unity3d.com/Manual/IL2CPP-BytecodeStripping.html).

**Comment by inventateq** [comment]
Solved by exporting my project as a unity package and importing it into a new project made in Unity 2018.3.14f1...after that you only have to write your tags again and delete the following files: androidx.browser.browser-1.0.0
androidx.browser.browser-1.0.0.aar
androidx.core.core-1.0.0
androidx.core.core-1.0.0.aar
from \YoureProjectLocation\Assets\Plugins\Android

got [aws training](https://www.inventateq.com/amazon-web-services-aws-training-institutes-bangalore.php) and [devops training](https://www.inventateq.com/devops-training-institutes-in-bangalore.php)by expert who has resolved and explained the above issue to me

**Comment by abiehusney25** [comment]
@inventateq @matteo-prosperi 

When I use the  Dll's that came with the UnitySDK [Found here](https://docs.aws.amazon.com/mobile/sdkforunity/developerguide/setup-unity.html), and with the one you added the Android build fails.
Failed to Build:
-     /gradleOut/unityLibrary/src/main/java/com/amazonaws/unity/Utils.java:11: error: cannot find symbol
-     location: class AWSUnityGCMWrapper
-     Temp/gradleOut/unityLibrary/src/main/java/com/amazonaws/unity/AWSUnityGCMWrapper.java uses or overrides a deprecated API.

This is for AWS SNS for Unity.
When I replace the dll's of sns,AWSSDK.SecurityToken, and some others, that I have found [here](https://aws.amazon.com/blogs/developer/referencing-the-aws-sdk-for-net-standard-2-0-from-unity-xamarin-or-uwp/) I get compile errors. They are the new netstandard2.0 dll's.

-     The name 'UnityInitializer' does not exist in the current context
-     Cannot convert lambda expression to type 'CancellationToken' because it is not a delegate type
-     Cannot convert lambda expression to type 'CancellationToken' because it is not a delegate type

For more information please see [here](https://github.com/aws/aws-sdk-net/issues/1515)
but if there is any updates on this please help. Thank you so much.

**Comment by QuentinGprd** [comment]
> Cannot convert lambda expression to type 'CancellationToken' because it is not a delegate

It's been a while since I've used the AWS SDK in Unity, but I think you no longer need to use `UnityInitializer` anymore. Just follow the fundamentals [here](https://github.com/aws/aws-sdk-net/blob/master/Unity.README.md).

As for `CancellationToken`, I guess you need to specify the namespace [System.Threading](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=netframework-4.8).

**Comment by abiehusney25** [comment]
> > Cannot convert lambda expression to type 'CancellationToken' because it is not a delegate
> 
> It's been a while since I've used the AWS SDK in Unity, but I think you no longer need to use `UnityInitializer` anymore. Just follow the fundamentals [here](https://github.com/aws/aws-sdk-net/blob/master/Unity.README.md).
> 
> As for `CancellationToken`, I guess you need to specify the namespace [System.Threading](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=netframework-4.8).

Thank you for the response, but I followed the instructions as seen and when I use the dll's given [here](https://github.com/aws/aws-sdk-net/blob/master/Unity.README.md) it cant build apks.

And when I use the updated dll's I get the compiler errors. 

Is there a newer sdk for Unity, if so how do I put it in?
See [here](https://github.com/aws/aws-sdk-net/issues/1515) for more info

**Comment by n8zach** [comment]
Why was this issue closed when there are only hacky work arounds and no fix!

**Comment by sporeservant** [comment]
Using the provided .zip of netstandard2 .dlls did not prove any more fruitful than using the SDK as unity packages. I also spent a lot of time fiddling with downloading packages manually from nuget and extracting DLLs into my Unity project, only to find a ton of version discrepancies.

Ultimately, the only thing that worked was stealing the Assets/ dir .dlls from this kind hero's example repo: https://github.com/rms80/awscognito-unity-accounts-demo

The fact that that folder is 2 years old and I still need it to operate in Unity (and I am using Unity 2020) says some attention needs to be paid to the Unity SDK. Ideally, there would be a unity AWS SDK for each major version of Unity, and a "latest" option for people trying to push the boundaries. Being that I am using a beta version of Unity, I expected to encounter some issues, but this really was quite a difficult operation even for an experienced developer.

**Comment by advantage-software** [comment]
Any updates to this issue?

**Comment by marcusice** [comment]
Still a lot of views about this issue.
You may watch this video to solve that step by step:
[https://youtu.be/uRNUJzQzNTk](https://youtu.be/uRNUJzQzNTk)

**Comment by AdamBebko** [comment]
Yes this is very difficult to set up for a noob like me. There were no clues that Unity Mobile SDK was deprecated until I happened on some forum posts looking for help. An up-to-date unity-centric tutorial provided by AWS would be really appreciated.

**Comment by karunateam** [comment]
Amazon, can you either formally state that you no longer support Unity with the latest AWS SDK for .Net v3+ or provide a solution for the missing UnityInitializer? Or at least explicitly declare whether UnityInitializer is required anymore or not?

Hello?

**Comment by josette-seitz** [comment]
I agree with @karunateam! Please provide a solution for missing UnityInitializer....

**Comment by BtreeSystemsIT** [comment]
To know more about AWS visit our site - <a href="https://www.btreesystems.com/aws-with-bigdata-training/">aws with bigdata training</a>

**Comment by natasha258** [comment]
> Solved by exporting my project as a unity package and importing it into a new project made in Unity 2018.3.14f1...after that you only have to write your tags again and delete the following files: androidx.browser.browser-1.0.0 androidx.browser.browser-1.0.0.aar androidx.core.core-1.0.0 androidx.core.core-1.0.0.aar from \YoureProjectLocation\Assets\Plugins\Android
> 
> got [aws training](https://www.inventateq.com/amazon-web-services-aws-training-institutes-bangalore.php) and [devops training](https://www.inventateq.com/devops-training-institutes-in-bangalore.php)by expert who has resolved and explained the above issue to me

just go through this <a href="https://intellimindz.com/devops-training-in-chennai/">devops training in chennai</a>
|| <a href="https://intellimindz.com/data-science-course-in-chennai/">data science course in chennai</a>

**Comment by sanyakishan** [comment]
Very Informative blog thank you for sharing. Keep sharing.
<a href="https://intellimindz.com/sas-training-in-chennai/">SAS Training in Chennai</a>

**Comment by santoshdevops321** [comment]
ohh thank you for this man. good job keep it upppp...
https://devopsacademy.co/
https://devopsacademy.co/devops-training/
https://devopsacademy.co/terraform-training-in-bangalore/
https://devopsacademy.co/aws-training-in-bangalore/
https://devopsacademy.co/top-5-devops-tools-in-2021/
https://devopsacademy.co/welcome-to-uni-education/
https://devopsacademy.co/about-us/

**Comment by sagarika-1** [comment]
Very informative blog, thanks for sharing. [Tosca Training in Bangalore](https://dturtleacademy.com/tosca-training-in-bangalore/)

**Comment by ganapathi6779-beep** [comment]
This suggests that the underlying DLL or type-resolution logic (e.g. NuGet-resolved assemblies) is incompatible with Unity 2019.1’s Android build process. It might be caused by Unity changing how AndroidJavaObject or array parameters are marshalled, or by a mismatch between the versions of the AWS SDK and Unity’s expected runtime.[ Prompt Engineering course in Hyderabad](https://promptacademy.in/ <a href="https://promptacademy.in/prompt-engineering-course-in-hyderabad/"prompt engineering course in Hyderabad)
